transformer_engine/jax/cpp_extensions.py

```diff
@@ -330,10 +330,23 @@ def lowering(ctx, x, gamma, beta, *, zero_centered_gamma, epsilon):
         batch_shape = out_shape[:-1]
         batch_size = reduce(operator.mul, x_shape) // hidden_size
 
+        # This is a dummy-call into the forward kernel to get workspace sizes
+        # where shape = meta[0] and dtype = meta[1]
+        work_meta, barrier_meta = transformer_engine_jax.get_layernorm_fwd_workspace_sizes(
+            batch_size,
+            hidden_size,
+            jax_dtype_to_te_dtype(x_aval.dtype),      # in te_dtype
+            jax_dtype_to_te_dtype(gamma_aval.dtype),  # weight te_dtype
+            jax_dtype_to_te_dtype(x_aval.dtype),      # out te_dtype (same as input for Fp16/Bf16)
+            True, zero_centered_gamma, epsilon
+        )
+
         out_types = [
             ir.RankedTensorType.get(out_shape, output_type),
             ir.RankedTensorType.get(batch_shape, ir_mu_dtype),
             ir.RankedTensorType.get(batch_shape, ir_rsigma_dtype),
+            ir.RankedTensorType.get(work_meta[0], te_dtype_to_ir_dtype(work_meta[1])),
```

Comment:
These changes make the number of outputs from `lowering` is different with `abstract`. I assume there would be a XLA assertion, could you test this to see if any potential issues?

Besides, these workspace tensors are never used, then it might  be removed during XLA compiling. We might have to test this also.

Reply:
The workspace tensors I added appear to be used in the C++ custom ops and NVTE kernels they call. They were being allocated with cudaMalloc in WorkspaceManager, which is what we want to replace with this PR. Did I misread/misunderstand something an about this?

It is entirely possible I unintentionally skipped over additional changes needed in the abstract (or elsewhere) to make this work correctly. If so, please let me know what's missing and I'll make sure it gets implemented. Thanks!

Reply:
Yes the purpose of this PR is to remove `cudaMalloc in WorkspaceManager`.

I am just thinking there might be two issues
1. The mismatching number of outputs between `lowering` and `abstract` might trigger some assertion. If so, then we have to seperate `abstract` for `inner_p` and `outer_p`, then add dummy outouts to `abstract` of inner_p.
2. Ununsed workspace tensors might be removed via some XLA's memory optimization pass. But after carefully thinking, we might not to worry about this, since XLA has less knowledge to custom calls, therefore it usually skips those optimization to custom calls

Reply:
1. If JAX doesn't raise an error due to different number of inputs/outputs, please raise this and give a repro (don't need to be a minimal one). I can look at adding the missing check in JAX.

2. XLA and all compilers can remove one full operation if all its outputs aren't used and is a side-effect free operation. But it can't remove only part of one operations outputs unless it is know that this instruction support this. So yes, XLA can't do that for custom_call, but in practice, it doesn't do that for native operations too . I don't know any such exception in XLA.